src/hclust.jl

```diff
@@ -316,6 +316,137 @@ function hclust2{T<:Real}(d::Symmetric{T}, method::Function)
     hcat(mr[o], mc[o]), h[o]
 end
 
+# Ward linkage algorithm, using a Lance-Williams formula.
+#
+# Based on the R-coded version of the Fortran function hclust.f, originally
+# written by Fionn Murtagh (1986) and modified for R by Ross Ihaka (1996), Fritz Leisch (2000)
+# and Martin Maechler (2001).
+#
+# As used in Murtagh, F. and P. Legendre,
+# “Ward’s hierarchical agglomerative clustering method: which algorithms implement Ward’s criterion?”
+# J. Classification 31(3) pp. 274--295, 2014
+#
+# currently supported methods: :ward1 and :ward2
+function hclust_lw{T<:Real}(ds::Symmetric{T}, method::Symbol)
+
+    nc = size(ds,1)                      # number of clusters
+    mr = Vector{Int64}(nc-1)               # min row
+    mc = Vector{Int64}(nc-1)               # min col
+    h = Vector{T}(nc-1)                  # height
+
+    triindex(n::Int64, i::Int64,j::Int64) = j + (i-1)*n - ((i*(i+1)) >> 1);
+
+    if method==:ward1
+        dscpy = [ds[r,c] for r=1:(nc-1) for c=(r+1):nc]
+    elseif method == :ward2
+        dscpy = [ds[r,c] for r=1:(nc-1) for c=(r+1):nc].^2
+    else
+        error("method ", method, " is unsupported by hclust_lw().")
+    end
+
+    merges = -collect(1:nc)
+    membr = ones(Float64, nc)
```

Comment:
`ones` and `zeros` default to `Float64`, no need to specific it (here and elsewhere).